### DISCUSSION TITLE
How to limit libp2p memory usage?

### DISCUSSION THREAD

**Discussion by Wiezzel** [libp2p/rust-libp2p#4572] [Q&A] [answered]
I am running a network of libp2p nodes, which use the following protocols:
```
    identify: identify::Behaviour,
    kademlia: Kademlia<MemoryStore>,
    autonat: Toggle<autonat::Behaviour>,
    relay: RelayClient,
    dcutr: dcutr::Behaviour,
    request: request_response::Behaviour<MessageCodec<T>>,
    gossipsub: gossipsub::Behaviour,
    ping: ping::Behaviour,
```
Notes:
 * `identify` is using interval 60s and pushing listen addr updates,
 * `kademlia` is bootstrapped every 5 minutes, and used for querying peers when addr is not known,
 * `relay` and `dcutr` are not actively used at the moment,
 * `gossipsub` is used for most communication - there is a single topic that every peer is subscribed to and everyone pushes one message every 10s,
 * `request_response` is used for additional messages, but I have noticed the problem even for nodes which don't send or receive any requests, so it can probably be ignored.

The problem which I have is that the process seems to be using too much memory. The usage grows ~15MB/sec seemingly without any limit (process gets killed by OOM at 10 GB). I've tried adding `libp2p_memory_connection_limits::Behaviour`, but it doesn't seem to help. Here is a flamegraph which I have generated using jemalloc. If I read it correctly (high chance I am not though), it's mostly `libp2p_kad` that's allocating memory.

![flame](https://github.com/libp2p/rust-libp2p/assets/9394821/9c3626cc-9816-42f7-87d5-6764bb8d6f1e)

How can I limit the memory usage of my node to some reasonable value (e.g. 4 GB)?

libp2p version: `0.52.3`

**Comment by mxinden** [comment]
Thank you for the report! Much appreciated, especially the framegraph. I will take a more in-depth look tomorrow.

If you can, in the meantime, are you able to share more code? Also what network topology do you have? Are you connecting to a known network, e.g. IPFS or Ethereum?

**Reply by Wiezzel** [reply]
That's a private repo, so I'm not sure if I can share more, but I will check. We're not connecting to any known network, such as IPFS or Ethereum, it's an isolated network. The network looks like this:
 * Two bootnodes with publicly known peer IDs and addresses which just run, not doing anything special.
 * One special node called *scheduler* which listens on gossipsub messages and also send and receives some requests.
 * Many *worker* nodes which connect to the bootnodes, regularly (every 10s) publish gossipsub messages and also exchange some requests with the *scheduler* and *clients*. They all have public addresses.
 * Many *client* nodes which connect to the bootnodes, listen for gossipsub messages and exchanges some requests with the *scheduler* and *workers*. The are not required to have public addresses and most of them probably don't.
All the nodes perform kademlia bootstrap every 5 minutes and start kademlia queries if they don't know the address of a node they want to send a request to.

**Reply by Wiezzel** [reply]
@mxinden [here](https://drive.google.com/file/d/1G2t4uV7sF6Fr0KoQXo2BHVaLUR9F5_MT/view?usp=drive_link) you can view interactive version of the flame graph (seems like GitHub butchered this one when uploading).

**Comment by thomaseizinger** [comment]
Can you do some investigation with `heaptrack` to see where the allocations are coming from? Do you have a custom `MemoryStore` config?

**Reply by Wiezzel** [reply]
>Can you do some investigation with heaptrack to see where the allocations are coming from?

Will do and share the results in a while.

>Do you have a custom MemoryStore config?

Nay, all default.
```rust
kademlia: Kademlia::with_config(
    local_peer_id,
    MemoryStore::new(local_peer_id),
    Default::default(),
)
```

**Reply by Wiezzel** [reply]
@thomaseizinger Heaptrack analysis shows that most leaked allocations are from `libp2p_gossipsub::behaviour::Behaviour::forward_msg`. You can check the heaptrack output [here](https://drive.google.com/file/d/1HKoeyzFWvpyqEz3zIh4SLrVyBN3WU7uW/view?usp=drive_link)

**Reply by mxinden** [reply]
Sharing thoughts early:

[`forward_msg`](https://github.com/libp2p/rust-libp2p/blob/master/protocols/gossipsub/src/behaviour.rs#L2698-L2771) distributes gossipsub messages to the various peers of the local node. For that they are first pushed into `Behaviour::events`, then emitted to the `Swarm`, which then sends them to the `ConnectionHandler`, which then stores them into `EnabledHandler::send_queue`.

https://github.com/libp2p/rust-libp2p/blob/ecdd0ff7677e59af1a1fa63de170e16f75a94f76/protocols/gossipsub/src/handler.rs#L99-L100

Off the top of my head I am assuming that you have one or more `ConnectionHandler`s operating on slow connections, not able to deliver the messages to their remote peers, i.e. not able to drain `EnabledHandler::send_queue` faster than it is filled up from `forward_msg`.

In case you have some time debugging, would you mind adding a `println!("{}", self.send_queue.size())` into `EnabledHandler::poll` @Wiezzel just to smoke test whether it is `send_queue` which is growing unboundedly?

**Reply by Wiezzel** [reply]
@mxinden 
>In case you have some time debugging, would you mind adding a `println!("{}", self.send_queue.size())` into `EnabledHandler::poll` just to smoke test whether it is `send_queue` which is growing unboundedly?

Your suspicions were right. After a few seconds of process running I see numbers like ~20k. Now, how to fix?

**Reply by Wiezzel** [reply]
There's also one more thing about the gossipsub messages. It seems like some of the delivered messages are many days old. That's something I would like to eliminate, because these messages are pretty much worthless for the application logic. All the messages which our *workers* broadcast are their current states. So if a newer message with the same publisher is available, we can drop all the older ones.

I have tried interpreting the messages' sequence numbers as timestamps (I think they're implemented as nanoseconds since UNIX epoch, right?), but then I occasionally see dates like `2513-04-28T03:23:56.028264293+00:00` which made me think this probably isn't a reliable mechanism.

Is there a way to manipulate gossipsub only to broadcast the most recent messages from each publisher?

**Reply by mxinden** [reply]
> Your suspicions were right. After a few seconds of process running I see numbers like ~20k. Now, how to fix?

Great that we found the culprit.

@Wiezzel can you share your GossipSub config here? Even better the entire code that instantiates your root `NetworkBehaviour`?

**Reply by Wiezzel** [reply]
@mxinden 
> can you share your GossipSub config here? Even better the entire code that instantiates your root NetworkBehaviour?

There you go. Nothing fancy in configs, I guess.
```rust
let mut request_config = request_response::Config::default();
request_config.set_request_timeout(Duration::from_secs(60));
let behaviour = Behaviour {
    relay,
    identify: identify::Behaviour::new(
        identify::Config::new(protocol, keypair.public())
            .with_interval(Duration::from_secs(60))
            .with_push_listen_addr_updates(true),
    ),
    kademlia: Kademlia::with_config(
        local_peer_id,
        MemoryStore::new(local_peer_id),
        Default::default(),
    ),
    autonat: Toggle::from(
        (!private_node).then(|| autonat::Behaviour::new(local_peer_id, Default::default())),
    ),
    dcutr: dcutr::Behaviour::new(local_peer_id),
    request: request_response::Behaviour::new(
        vec![(WORKER_PROTOCOL, ProtocolSupport::Full)],
        request_config,
    ),
    gossipsub: gossipsub::Behaviour::new(
        MessageAuthenticity::Signed(keypair),
        Default::default(),
    )
    .unwrap(),
    ping: ping::Behaviour::new(Default::default()),
};
```

**Reply by Wiezzel** [reply]
Additional information (might be important for tracking down the issue): I see quite a lot of these warnings in the logs:
```
[2023-09-29T12:00:43Z WARN  libp2p_gossipsub::handler] The maximum number of inbound substreams attempts has been exceeded
[2023-09-29T12:00:59Z WARN  libp2p_gossipsub::handler] The maximum number of outbound substream attempts has been exceeded
```

**Reply by mxinden** [reply]
@Wiezzel what implementations and implementation versions are being used? E.g. only rust-libp2p? Only rust-libp2p v0.52?

**Reply by Wiezzel** [reply]
@mxinden Only `rust-libp2p`. Current nodes have `v0.52`, but it is possible that there are some old ones which are still running `v0.50` (we don't have control over them, it's an open network).

**Reply by thomaseizinger** [reply]
@mxinden So this is a backpressure issue, correct? We shouldn't overload a particular `ConnectionHandler` if it is struggleing to process our messages.

**Reply by Wiezzel** [reply]
@thomaseizinger Yes, that seems to be the problem. The `send_queue` in gossipsub's `EnabledHandler` is unbounded and when it receives behaviour events at a higher rate than it's able to send, this queue grows indefinitely.

What's still to find out is why in my case the messages were flowing in so fast, but couldn't be sent.

**Reply by mxinden** [reply, marked as answer]
I see two solutions:

1. Front or tail drop in `EnabledHandler` once `send_queue` reaches a certain size.
2. Implement backpressure between Gossipsubs `NetworkBehaviour` and `ConnectionHandler` and only forward from `NetworkBehaviour` to `ConnectionHandler` when the latter can handle another message. I.e. drop messages for a specific peer in `NetworkBehaviour` instead of as suggested in (1) in `ConnectionHandler`.

**Comment by mxinden** [comment]
@Wiezzel after our call last Friday, any updates? Where you able to find the source for the intermittent bug, i.e. the old messages?

**Reply by Wiezzel** [reply]
@mxinden Sorry, I prefer not to work on weekends. 😛 But I will investigate this further.

**Reply by Wiezzel** [reply]
@mxinden I've tried a couple of times to recreate this with a very small network where some nodes run libp2p `0.50` and some run `0.52`, but without any luck. Maybe I would have to add some nodes with very slow connection to fully reflect the original circumstances, but I don't see any easy way to do this.
